apis/v1alpha1/tcxProgram_types.go

```diff
@@ -0,0 +1,93 @@
+/*
+Copyright 2024.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// All fields are required unless explicitly marked optional
+// +kubebuilder:validation:Required
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +genclient:nonNamespaced
+//+kubebuilder:object:root=true
+//+kubebuilder:subresource:status
+//+kubebuilder:resource:scope=Cluster
+
+// TcxProgram is the Schema for the TcxProgram API
+// +kubebuilder:printcolumn:name="BpfFunctionName",type=string,JSONPath=`.spec.bpffunctionname`
+// +kubebuilder:printcolumn:name="NodeSelector",type=string,JSONPath=`.spec.nodeselector`
+// +kubebuilder:printcolumn:name="Status",type=string,JSONPath=`.status.conditions[0].reason`
+// +kubebuilder:printcolumn:name="Direction",type=string,JSONPath=`.spec.direction`,priority=1
+// +kubebuilder:printcolumn:name="InterfaceSelector",type=string,JSONPath=`.spec.interfaceselector`,priority=1
+// +kubebuilder:printcolumn:name="Position",type=string,JSONPath=`.spec.position`,priority=1
+type TcxProgram struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec TcxProgramSpec `json:"spec"`
+	// +optional
+	Status TcxProgramStatus `json:"status,omitempty"`
+}
+
+// TcxProgramSpec defines the desired state of TcxProgram
+type TcxProgramSpec struct {
+	TcxProgramInfo `json:",inline"`
+	BpfAppCommon   `json:",inline"`
+}
+
+// TCXPositionType defines the TCX position string
+type TCXPositionType string
+
+const (
+	// TCXFirst TCX program will be the first program to execute.
+	TCXFirst TCXPositionType = "First"
+
+	// TCXLast TCX program will be the last program to execute.
+	TCXLast TCXPositionType = "Last"
+)
+
+// TcxProgramInfo defines the tc program details
+type TcxProgramInfo struct {
```

Comment:
We should also include revision.

Reply:
yes I can add it and TBD how to make sense of it

Reply:
I think it works like it's supposed to work.  We're just not sure how we would use it yet.

Reply:
I don't think using pri for tcx is right API given how cilium and libbpf apis look like

Reply:
I'm for bpfman, itself, supporting an API that is similar to the ones supported by cilium and libbpf, but I don't think this API works well for a Kubernetes cluster.